### PR TITLE
 Don't interleave handshake and other record types in TLSv1.3

### LINKED ISSUE #8189: Fragmented handshake messages with app data between them are not rejected
Using 1980ce45d6b

When OpenSSL receives handshake protocol messages that are fragmented over multiple records split by application data, the connection is not aborted.

This is a [RFC 8446 section 5.1](https://tools.ietf.org/html/rfc8446#section-5.1) violation:
```
   -  Handshake messages MUST NOT be interleaved with other record
      types.  That is, if a handshake message is split over two or more
      records, there MUST NOT be any other records between them.
```

Interestingly, when the message is fragmented in a 4 byte and 1 byte record, the server exhibits correct behaviour.

There's also a minor issue with parsing the message: when it is too large (2 instead of 1 byte), the server replies with `illegal_parameter` rather than the expected `decode_error`.

   <details><summary>reproducer</summary><p>

```
openssl req -x509 -newkey rsa -keyout localhost.key -out localhost.crt -subj /CN=localhost -nodes -batch
openssl s_server -key localhost.key -cert localhost.crt -www 2>server.err >server.out &
openssl_pid=$!
git clone https://github.com/tomato42/tlsfuzzer
pushd tlsfuzzer
# won't be needed after https://github.com/tomato42/tlsfuzzer/pull/501 is merged
git checkout key-update-handshake
git clone https://github.com/tomato42/tlslite-ng .tlslite-ng
pushd .tlslite-ng
# won't be needed after https://github.com/tomato42/tlslite-ng/pull/341 is merged
git checkout key-update-support
popd
ln -s .tlslite-ng/tlslite tlslite
git clone https://github.com/warner/python-ecdsa .python-ecdsa
ln -s .python-ecdsa/src/ecdsa ecdsa
PYTHONPATH=. python scripts/test-tls13-keyupdate.py -e 'app data split, conversation with KeyUpdate msg'
popd
kill $openssl_pid
```
   </p></details>

   <details><summary>OpenSSL output</summary><p>

```
Using default temp DH parameters
ACCEPT
...
140131671496512:error:14160098:SSL routines:read_state_machine:excessive message size:ssl/statem/statem.c:601:
...
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:
140131671496512:error:1420607A:SSL routines:tls_process_key_update:bad key update:ssl/statem/statem_lib.c:648:

```
   </p></details>

   <details><summary>tlsfuzzer output</summary><p>

```
large KeyUpdate message ...
Error encountered while processing node <tlsfuzzer.expect.ExpectAlert object at 0x7f39b52a6150> (child: <tlsfuzzer.expect.ExpectClose object at 0x7f39b52a6210>) with last message being: <tlslite.messages.Message object at 0x7f39b5069c90>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-keyupdate.py", line 575, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 223, in run
    node.process(self.state, msg)
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/expect.py", line 1366, in process
    raise AssertionError(problem_desc)
AssertionError: Expected alert description "decode_error" does not match received "illegal_parameter"

...

3/2 fragmented keyupdate msg, appdata between ...
Error encountered while processing node ExpectNewSessionTicket() (child: ExpectNewSessionTicket()) with last message being: <tlslite.messages.Message object at 0x7f39b4d76b90>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-keyupdate.py", line 575, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: ApplicationData(len=3863)

...

2/3 fragmented keyupdate msg, appdata between ...
Error encountered while processing node ExpectNewSessionTicket() (child: ExpectNewSessionTicket()) with last message being: <tlslite.messages.Message object at 0x7f39b4d7be10>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-keyupdate.py", line 575, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 221, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: ApplicationData(len=3863)

Test with KeyUpdate msg with different msg_type or fragmented msg.
Verify that server will correctly handle updating the keys
or refuse the connection with relevant Alert msg.
version: 1

Test end
successful: 262
failed: 4
  '1/4 fragmented keyupdate msg, appdata between'
  '2/3 fragmented keyupdate msg, appdata between'
  '3/2 fragmented keyupdate msg, appdata between'
  'large KeyUpdate message'
```
   </p></details>

See also #8188 for 'app data split, conversation with KeyUpdate msg' test case

### PR DESCRIPTION
**Pull request by mattcaswell**
In TLSv1.3 it is illegal to interleave handshake records with non handshake records.

Fixes #8189

### CODE REVIEW

**Comment by tomato42** [comment]
fixes the interleaving issue, the wrong and/or inconsistent alert remains

**Comment by mattcaswell** [comment]
> fixes the interleaving issue, the wrong and/or inconsistent alert remains

As noted in #8189 I don't plan to make a change because of that.

**Comment by mattcaswell** [comment]
Ping?

File: test/recipes/70-test_sslrecords.t [hunk 1]
```diff
@@ -207,9 +208,24 @@ SKIP: {
     $fatal_alert = 0;
     $proxy->clear();
     $boundary_test_type = DATA_AFTER_KEY_UPDATE;
-    $proxy->filter(\&not_on_record_boundary);
```
Comment (kaduk) [inline test/recipes/70-test_sslrecords.t]: Are these ```$proxy->filter``` assignment removals just unrelated cleanup or were they functional changes?

Comment (mattcaswell) [reply test/recipes/70-test_sslrecords.t]: Unrelated clean up. The filter doesn't get cleared by a `$proxy->clear()` call so there is no need to set them every time.

Comment (kaduk) [reply test/recipes/70-test_sslrecords.t]: That's what it looked like, but thanks for confirming.

**Comment by mattcaswell** [comment]
Pushed. Thanks.
